File: src/08_Amortized_Analysis/subarray_sum.cpp
#include <bits/stdc++.h>

using namespace std;


vector<int> v{1, 3, 2, 5, 1, 1, 2, 3};
const int n = 8;
const int x = 8;


// amortized O(n)
// If a value of the array is >= x, this alg gets stuck.
void solve()
{
  size_t l = 0, r = 0;
  size_t sum = v[0];

  while ((r >= l) && (r < n)) {
    if (sum == x) {
      copy(v.begin() + l, v.begin() + r + 1, ostream_iterator<int>(cout, " "));
      cout << '\n';
      return;
    }
    if ((sum < x) && (sum + v[r + 1] <= x)) {
      ++r;
      sum += v[r];
    } else {
      sum -= v[l];
      ++l;
    }
  }
  cout << "IMPOSSIBLE\n";
}


// amortized O(n)
void solve2()
{
  int l = 0;
  int r = 0;
  int s = v[0];
  while (r < n) {
    if (s == x) {
      copy(v.begin() + l, v.begin() + r + 1, ostream_iterator<int>(cout, " "));
      cout << '\n';
      return;
    }
    if (s < x) {
      if (r == n - 1) break;
      ++r;
      s += v[r];
    }
    else {
      s -= v[l];
      if (l == r) ++r;
      ++l;
    }
  }
  cout << "IMPOSSIBLE\n";
}


int main() {solve2();}

File: src/08_Amortized_Analysis/subarray_sum.py
v = [1, 3, 2, 5, 1, 1, 2, 3]


# amortized O(n)
def solve(x):
  n = len(v)
  left = 0
  right = 0
  s = v[0]

  while right < n:
    if s == x:
      print(f"{x}: {v[left: right + 1]}")
      return

    if s < x:
      if right == n - 1:
        break
      right += 1
      s += v[right]
    else:
      if left == right:
        if right == n - 1:
          break
        right += 1
        s += v[right]
      s -= v[left]
      left += 1

  print(f"{x}: IMPOSSIBLE")


if __name__ == '__main__':
  solve(8)
  solve(10)
  solve(16)
  solve(6)
  solve(12)
  solve(4)
  solve(24)
